### PR TITLE
fix and test empty CFTimeIndex

### LINKED ISSUE #7298: html repr fails for empty cftime arrays
### What happened?

The html repr of a cftime array wants to display the "calendar", which it cannot if it is empty.

### What did you expect to happen?

No error.

### Minimal Complete Verifiable Example

```Python
import numpy as np
import xarray as xr


data_obs = np.random.randn(3)
time_obs = xr.date_range("2000-01-01", periods=3, freq="YS", calendar="noleap")

obs = xr.DataArray(data_obs, coords={"time": time_obs})

o = obs[:0]

xr.core.formatting_html.array_repr(o)
```

### MVCE confirmation

- [ ] Minimal example — the example is as focused as reasonably possible to demonstrate the underlying issue in xarray.
- [ ] Complete example — the example is self-contained, including all data and the text of any traceback.
- [ ] Verifiable example — the example copy & pastes into an IPython prompt or [Binder notebook](https://mybinder.org/v2/gh/pydata/xarray/main?urlpath=lab/tree/doc/examples/blank_template.ipynb), returning the result.
- [ ] New issue — a search of GitHub Issues suggests this is not a duplicate.

### Relevant log output

```Python
ValueError                                Traceback (most recent call last)
Input In [1], in <cell line: 12>()
      8 obs = xr.DataArray(data_obs, coords={"time": time_obs})
     10 o = obs[:0]
---> 12 xr.core.formatting_html.array_repr(o)

File ~/code/xarray/xarray/core/formatting_html.py:318, in array_repr(arr)
    316 if hasattr(arr, "xindexes"):
    317     indexes = _get_indexes_dict(arr.xindexes)
--> 318     sections.append(index_section(indexes))
    320 sections.append(attr_section(arr.attrs))
    322 return _obj_repr(arr, header_components, sections)

File ~/code/xarray/xarray/core/formatting_html.py:195, in _mapping_section(mapping, name, details_func, max_items_collapse, expand_option_name, enabled)
    188 expanded = _get_boolean_with_default(
    189     expand_option_name, n_items < max_items_collapse
    190 )
    191 collapsed = not expanded
    193 return collapsible_section(
    194     name,
--> 195     details=details_func(mapping),
    196     n_items=n_items,
    197     enabled=enabled,
    198     collapsed=collapsed,
    199 )

File ~/code/xarray/xarray/core/formatting_html.py:155, in summarize_indexes(indexes)
    154 def summarize_indexes(indexes):
--> 155     indexes_li = "".join(
    156         f"<li class='xr-var-item'>{summarize_index(v, i)}</li>"
    157         for v, i in indexes.items()
    158     )
    159     return f"<ul class='xr-var-list'>{indexes_li}</ul>"

File ~/code/xarray/xarray/core/formatting_html.py:156, in <genexpr>(.0)
    154 def summarize_indexes(indexes):
    155     indexes_li = "".join(
--> 156         f"<li class='xr-var-item'>{summarize_index(v, i)}</li>"
    157         for v, i in indexes.items()
    158     )
    159     return f"<ul class='xr-var-list'>{indexes_li}</ul>"

File ~/code/xarray/xarray/core/formatting_html.py:140, in summarize_index(coord_names, index)
    138 index_id = f"index-{uuid.uuid4()}"
    139 preview = escape(inline_index_repr(index))
--> 140 details = short_index_repr_html(index)
    142 data_icon = _icon("icon-database")
    144 return (
    145     f"<div class='xr-index-name'><div>{name}</div></div>"
    146     f"<div class='xr-index-preview'>{preview}</div>"
   (...)
    150     f"<div class='xr-index-data'>{details}</div>"
    151 )

File ~/code/xarray/xarray/core/formatting_html.py:132, in short_index_repr_html(index)
    129 if hasattr(index, "_repr_html_"):
    130     return index._repr_html_()
--> 132 return f"<pre>{escape(repr(index))}</pre>"

File ~/code/xarray/xarray/core/indexes.py:547, in PandasIndex.__repr__(self)
    546 def __repr__(self):
--> 547     return f"PandasIndex({repr(self.index)})"

File ~/code/xarray/xarray/coding/cftimeindex.py:353, in CFTimeIndex.__repr__(self)
    345     end_str = format_times(
    346         self.values[-REPR_ELLIPSIS_SHOW_ITEMS_FRONT_END:],
    347         display_width,
    348         offset=offset,
    349         first_row_offset=offset,
    350     )
    351     datastr = "\n".join([front_str, f"{' '*offset}...", end_str])
--> 353 attrs_str = format_attrs(self)
    354 # oneliner only if smaller than display_width
    355 full_repr_str = f"{klass_name}([{datastr}], {attrs_str})"

File ~/code/xarray/xarray/coding/cftimeindex.py:272, in format_attrs(index, separator)
    267 def format_attrs(index, separator=", "):
    268     """Format attributes of CFTimeIndex for __repr__."""
    269     attrs = {
    270         "dtype": f"'{index.dtype}'",
    271         "length": f"{len(index)}",
--> 272         "calendar": f"'{index.calendar}'",
    273         "freq": f"'{index.freq}'" if len(index) >= 3 else None,
    274     }
    276     attrs_str = [f"{k}={v}" for k, v in attrs.items()]
    277     attrs_str = f"{separator}".join(attrs_str)

File ~/code/xarray/xarray/coding/cftimeindex.py:698, in CFTimeIndex.calendar(self)
    695 """The calendar used by the datetimes in the index."""
    696 from .times import infer_calendar_name
--> 698 return infer_calendar_name(self)

File ~/code/xarray/xarray/coding/times.py:374, in infer_calendar_name(dates)
    371             return sample.calendar
    373 # Error raise if dtype is neither datetime or "O", if cftime is not importable, and if element of 'O' dtype is not cftime.
--> 374 raise ValueError("Array does not contain datetime objects.")

ValueError: Array does not contain datetime objects.
```

### Anything else we need to know?

Bisected to 7379923de756a2bcc59044d548f8ab7a68b91d4e use `_repr_inline_` for indexes that define it.

### Environment

<details>

</details>

### PR DESCRIPTION
**Pull request by mathause**
- [x] Closes #7298
- [x] Tests added
- [ ] User visible changes (including notable bug fixes) are documented in `whats-new.rst`
- [ ] New functions/methods are listed in `api.rst`

Otherwise `da.indexes` and the html repr raise a `ValueError`. I first had `"<undefined>"` but I think `None` is better. cc @spencerkclark @keewis

### CODE REVIEW

File: xarray/coding/cftimeindex.py [hunk 1]
```diff
@@ -272,7 +272,7 @@ def format_attrs(index, separator=", "):
     attrs = {
         "dtype": f"'{index.dtype}'",
         "length": f"{len(index)}",
-        "calendar": f"'{index.calendar}'",
+        "calendar": f"'{index.calendar}'" if len(index) else None,
```
[the remark this answers is not in the document]
Comment (mathause) [reply xarray/coding/cftimeindex.py:275]: Yes that will solve the problem on a lower level. However, empty `CFTimeIndex` also fail for other operations:

```python
import xarray as xr
index = xr.coding.cftimeindex.CFTimeIndex([])
index.asi8 # errors
index.freq # errors
index.calendar # errors
```

I think we should either fix them all or keep the current fix. @spencerkclark: do you have an opinion?

[the remark this answers is not in the document]
Comment (spencerkclark) [reply xarray/coding/cftimeindex.py:275]: I agree, I think it makes sense to fix all of those.  Sorry for not chiming in on the original issue.  Indeed at the moment I'm not sure if caching the calendar would provide that much value (and we'd still have to deal with the edge case that one constructed an empty `CFTimeIndex` from the start), so I think it's fine to return `None`.

**Comment by mathause** [comment]
I updated `CFTimeIndex` to work for no data. I tried to get the same behavior as pandas and am happy to get feedback.

File: xarray/coding/cftimeindex.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mathause) [reply xarray/coding/cftimeindex.py:275]: This change is still required to have the quotes for string calendar/ freq and not have them for None. (But fixing `CFTimeIndex` is a good idea nonetheless.)

**Review by spencerkclark — COMMENTED** [review]
Thanks @mathause!  This looks pretty much ready for a what's new entry from my perspective.

File: xarray/coding/cftimeindex.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mathause) [reply xarray/coding/cftimeindex.py:275]: I found that `f"{index.calendar!r}" does what we want - so going for this.

**Review by spencerkclark — APPROVED** [review]
Looks good — thanks again!

**Pull request merged**